homeassistant/components/totalconnect/config_flow.py

```diff
@@ -44,13 +57,93 @@ async def async_step_user(self, user_input=None):
+                if self.usercodes[location_id] is None:
+                    valid = await self.hass.async_add_executor_job(
+                        self.client.locations[location_id].set_usercode,
+                        user_entry[CONF_LOCATION],
+                    )
+                    if valid:
+                        self.usercodes[location_id] = user_entry[CONF_LOCATION]
+                    else:
+                        errors[CONF_LOCATION] = "usercode"
+                    break
+
+            complete = True
+            for location_id in self.usercodes:
+                if self.usercodes[location_id] is None:
+                    complete = False
+
+            if not errors and complete:
+                return self.async_create_entry(
+                    title="Total Connect",
+                    data={
+                        CONF_USERNAME: self.username,
+                        CONF_PASSWORD: self.password,
+                        CONF_USERCODES: self.usercodes,
+                    },
+                )
+        else:
+            for location_id in self.client.locations:
+                self.usercodes[location_id] = None
+
+        # show the next location that needs a usercode
+        location_codes = {}
+        for location_id in self.usercodes:
+            if self.usercodes[location_id] is None:
+                location_codes[
+                    vol.Required(
+                        CONF_LOCATION,
+                        default=location_id,
+                    )
+                ] = str
+                break
+
+        data_schema = vol.Schema(location_codes)
+        return self.async_show_form(
+            step_id="locations",
+            data_schema=data_schema,
+            errors=errors,
+            description_placeholders={"base": "description"},
+        )
+
+    async def async_step_reauth(self, config):
+        """Perform reauth upon an authentication error or no usercode."""
+        self.username = config[CONF_USERNAME]
+        self.usercodes = config[CONF_USERCODES]
+
+        return await self.async_step_reauth_confirm()
+
+    async def async_step_reauth_confirm(self, user_input=None):
+        """Dialog that informs the user that reauth is required."""
+        if user_input is None:
+            return self.async_show_form(
+                step_id="reauth_confirm",
+                data_schema=PASSWORD_DATA_SCHEMA,
+            )
 
-    async def is_valid(self, username="", password=""):
-        """Return true if the given username and password are valid."""
         client = await self.hass.async_add_executor_job(
-            TotalConnectClient.TotalConnectClient, username, password
+            TotalConnectClient.TotalConnectClient,
+            self.username,
+            user_input[CONF_PASSWORD],
+            self.usercodes,
         )
-        return client.is_valid_credentials()
+
+        if not client.is_valid_credentials():
```

Comment:
Please test this case. We need to show an error in the form.

Reply:
Test added.